modules/javafx.swing/src/main/java/javafx/embed/swing/InputMethodSupport.java

```diff
@@ -102,9 +122,14 @@ public AttributedCharacterIterator cancelLatestCommittedText(AttributedCharacter
 
         @Override
         public AttributedCharacterIterator getSelectedText(AttributedCharacterIterator.Attribute[] attributes) {
-            String text = fxRequests.getSelectedText();
-            if (text == null) text = "";
-            return new AttributedString(text).getIterator();
+            String[] selected = { null };
+            if (fxRequests != null) {
+                PlatformImpl.runAndWait(() -> {
+                    selected[0] = fxRequests.getSelectedText();
+                });
+            }
+            if (selected[0] == null) selected[0] = "";
+            return new AttributedString(selected[0]).getIterator();
```

Comment:
```
AtomicReference<String> ref = ...
...
String s = ref.get();
return new AttributedString(s == null ? "" : s).getIterator();
```

here and L102?

Reply:
I'm not familiar enough with AtomicReference to have an opinion on this. Someone else will have to weigh in on this.

Reply:
Either an array or an AtomicReference is OK in this case. AtomicReference is thread-safe, but that's not a concern here, since using `runAndWait` ensures that writing to the array happens before reading it.

Reply:
I might be wrong, but array would need to store its size (an int, 8 bytes on 64 bit machines) in addition to the pointers themselves, while AtomicReference needs just the pointer.
Plus, its semantics is much cleaner.

Reply:
Again, thread-safety isn't an issue here. Even if the OS were to invoke the IM methods on different threads (which is a terrifying thought) these adapters are created on a per-call basis so each thread would get its own instance. This also means those extra eight bytes are transitory.

If there's an established JavaFX pattern of using AtomicReference instead of an array I will gladly make the changes. But this does require manual testing (I still haven't figured out how to automate this) so I'm reluctant to do this based on personal style preferences.

Reply:
Here is more info [0].

We **are** dealing with multiple threads.

I've seen weird cases where modifying memory from different threads without explicit synchronization resulted in subtle bugs - we don't know all the details of cache coherence [1] implementation in multi-core systems, so I would strongly suggest to use AtomicReference.

[0] https://stackoverflow.com/questions/70381029/java-atomincinteger-vs-one-element-array-for-streams

[1] https://en.wikipedia.org/wiki/Cache_coherence

Reply:
Sorry, I was still thinking about the original thread-safety concern which involved simultaneous access to the fields in this adapter object. Not the issue here.

I agree that explicit synchronization needs to happen but I thought that was what `runAndWait` was for. @kevinrushforth also states that `runAndWait` should be sufficient. If you think there's a cache coherence problem that's much, much bigger than this PR since this pattern is used throughout the code base.

Reply:
> this pattern is used throughout the code base.

that might be the case, but I think we should use the right pattern for the new code.
single-element array is an anachronism, especially in the case of multi-threaded code as here.